torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp

```diff
@@ -14,6 +14,39 @@ namespace cuda {
 
 namespace {
 
+void insertMagicZero(
+    const std::vector<kir::ForLoop*>& loops,
+    const std::vector<IterDomain*>& loop_domains,
+    std::unordered_map<IterDomain*, Val*>& concrete_loop_idx_map) {
+  // Find magic zero insertion point
+  IterDomain* magic_zero_loop = nullptr;
+
+  // Search for proper magic zero insertion point,
+  //  starting from innermost.
+  for (auto idx : c10::irange(loops.size())) {
+    auto loop = loops[idx];
+    auto concrete_loop_id = GpuLower::current()->caMap()->getConcreteMappedID(
+        loop_domains[idx], IdMappingMode::EXACT);
+    auto loop_ind = concrete_loop_idx_map.at(concrete_loop_id);
+
+    // Save the concrete id if this loop id is decided to
+    //  be the insertion point by the magic zero util.
+    if (Index::protectWithMagicZero(loop, concrete_loop_id, loop_ind)) {
+      magic_zero_loop = concrete_loop_id;
+    }
+  }
+
+  // Insert magic zero if insertion point found
+  if (magic_zero_loop != nullptr &&
+      concrete_loop_idx_map.count(magic_zero_loop)) {
+    auto& ind = concrete_loop_idx_map.at(magic_zero_loop);
+    if (!ind->isConstScalar()) {
```

Comment:
Should this check be part of trying to find an insertion point? If an inner loop is a const scalar index, but a loop in a more outer position wasn't, shouldn't we just put it on the loop that's in a more outer position?

Reply:
I would presume when we have more optimal index hoisting we shouldn't need magic zero anymore 🤞

Reply:
Would need to revisit magic zeros in follow ups.